climpred/classes.py

```diff
@@ -84,20 +88,124 @@ class PredictionEnsemble:
     should house functions that both ensemble types can use.
     """
 
+    # ---------------
+    # Magic Functions
+    # ---------------
     @is_xarray(1)
     def __init__(self, xobj):
         if isinstance(xobj, xr.DataArray):
             # makes applying prediction functions easier, etc.
             xobj = xobj.to_dataset()
         has_dims(xobj, ['init', 'lead'], 'PredictionEnsemble')
-        self.initialized = xobj
-        self.uninitialized = {}
+        self._datasets = {'initialized': xobj}
+        # Reserve sub-dictionary for an uninitialized run.
+        self._datasets.update({'uninitialized': {}})
 
     # when you just print it interactively
     # https://stackoverflow.com/questions/1535327/how-to-print-objects-of-class-using-print
     def __repr__(self):
         return _display_metadata(self)
 
+    def __getattr__(self, name):
+        """Allows for xarray methods to be applied to our prediction objects.
+
+        Args:
+            * name: Function, e.g., .isel() or .sum().
+        """
+        # Temporarily registers attribute with the object.
+        setattr(self, name, self)
+
+        def wrapper(*args, **kwargs):
+            """Applies arbitrary function to all datasets in the PredictionEnsemble
+            object.
+
+            Got this from: https://stackoverflow.com/questions/41919499/
+            how-to-call-undefined-methods-sequentially-in-python-class
+            """
+
+            def _apply_func(v, name, *args, **kwargs):
+                """Handles exceptions in our dictionary comprehension.
+
+                In other words, this will skip applying the arbitrary function
+                to a sub-dataset if a ValueError is thrown. This specifically
+                targets cases where certain datasets don't have the given
+                dim that's being called. E.g., ``.isel(lead=0)`` should only
+                be applied to the initialized dataset.
+
+                Ref: https://stackoverflow.com/questions/1528237/
+                how-to-handle-exceptions-in-a-list-comprehensions
+                """
+                try:
+                    return getattr(v, name)(*args, **kwargs)
+                # This accounts for cases such as .sum(dim='time'). This doesn't
+                # apply it to the given dataset if the dimension doesn't exist.
+                except ValueError:
+                    return v
+                # This accounts for our custom error when applying some stats
+                # functions.
+                except DimensionError:
+                    return v
+
+            # Create temporary copy to modify to avoid inplace operation.
+            datasets = self._datasets.copy()
```

Comment:
This worries me a bit with big datasets. Not really sure what "copy" actually does, but if it does what it sounds like, if we have a 16 GB dataset, it'll duplicate that 16 GB dataset making memory usage == 32 GB, but I don't have another solution

Reply:
@ahuang11, I think we discussed that since `xarray` does this, there probably isn't a clear alternative to avoid inplace?

Reply:
Probably. Don't really know the inner workings of xarray :D